src/hotspot/share/gc/shenandoah/heuristics/shenandoahHeuristics.cpp

```diff
@@ -304,6 +436,155 @@ double ShenandoahHeuristics::time_since_last_gc() const {
+}
+
+void ShenandoahHeuristics::prepare_for_old_collections() {
+  assert(_generation->generation_mode() == OLD, "This service only available for old-gc heuristics");
+  ShenandoahHeap* heap = ShenandoahHeap::heap();
+  uint free_regions = 0;
+  size_t cand_idx = 0;
+  size_t total_garbage = 0;
+  size_t num_regions = heap->num_regions();
+
+  RegionData* candidates = _region_data;
+  for (size_t i = 0; i < num_regions; i++) {
+    ShenandoahHeapRegion* region = heap->get_region(i);
+    if (!in_generation(region))
+      continue;
+    else {
+      size_t garbage = region->garbage();
+      total_garbage += garbage;
+
+      candidates[cand_idx]._region = region;
+      candidates[cand_idx]._garbage = garbage;
+      cand_idx++;
+    }
+  }
+
+  // Give special treatment to humongous regions.  Assume humongous regions is entirely
+  // garbage or entirely non-garbage.  Assume that a head humongous region and the associated
+  // humongous continuous regions are uniformly entirely garbage or entirely non-garbage.
+  //
+  // Sift garbage humongous regions to front, non-garbage humongous regions to end of array.
+  size_t first_non_humongous_empty = 0;
+  size_t first_humongous_non_empty = cand_idx;
+
+  // This loop is written as while rather than for because of
+  // suspected gcc error in translating/optimizing for-loop
+  size_t i = 0;
+  while (i < first_humongous_non_empty) {
+    ShenandoahHeapRegion* region = candidates[i]._region;
+    if (region->is_humongous()) {
+      if (region->get_live_data_bytes() == 0) {
+        // Humongous region is entirely garbage.  Reclaim it.
+        if (i == first_non_humongous_empty) {
+          first_non_humongous_empty++;
+        } else {
+          RegionData swap_tmp = candidates[i];
+          candidates[i] = candidates[first_non_humongous_empty];
+          candidates[first_non_humongous_empty++] = swap_tmp;
+        }
+        i++;
+      } else {
+        // Humongous region is non garbage.  Don't reclaim it.
+        if (i + 1 == first_humongous_non_empty) {
+          first_humongous_non_empty--;
+          i++;
+        } else {
+          RegionData swap_tmp = candidates[i];
+          candidates[i] = candidates[--first_humongous_non_empty];
+          candidates[first_humongous_non_empty] = swap_tmp;
+          // Do not increment i so we can revisit swapped entry on next iteration
+        }
+      }
+    } else {
+      i++;
+    }
+  }
+
+
+  // Prioritize regions to select garbage-first regions
+  QuickSort::sort<RegionData>(candidates + first_non_humongous_empty, (int)(first_humongous_non_empty - first_non_humongous_empty),
+                              compare_by_garbage, false);
+
+  // Any old-gen region that contains 50% garbage or more is to be
+  // evacuated.  In the future, this threshold percentage may be specified on
+  // the command line or preferrably determined by dynamic heuristics.
+#define CollectionThresholdGarbagePercent 50
```

Comment:
We already have ShenandoahGarbageThreshold (defaults to 60 iirc) maybe we want to use this? Or use separate global/old/young thresholds?

Reply:
I've made and tested this change now.  Appears to me that default of ShenandoahGarbageThreshold is 25.  On established stress workloads, this change has the effect of choosing larger collection sets for old-gen.